src/module.c

```diff
@@ -1269,24 +1269,26 @@ static void module_import_(jl_task_t *ct, jl_module_t *to, jl_module_t *from, jl
     JL_UNLOCK(&world_counter_lock);
 }
 
-JL_DLLEXPORT void jl_module_import(jl_task_t *ct, jl_module_t *to, jl_module_t *from, jl_sym_t *s)
+JL_DLLEXPORT void jl_import_module(jl_task_t *ct, jl_module_t *JL_NONNULL m, jl_module_t *import, jl_sym_t *asname)
 {
-    module_import_(ct, to, from, s, s, 1);
-}
-
-JL_DLLEXPORT void jl_module_import_as(jl_task_t *ct, jl_module_t *to, jl_module_t *from, jl_sym_t *s, jl_sym_t *asname)
-{
-    module_import_(ct, to, from, asname, s, 1);
-}
-
-JL_DLLEXPORT void jl_module_use(jl_task_t *ct, jl_module_t *to, jl_module_t *from, jl_sym_t *s)
-{
-    module_import_(ct, to, from, s, s, 0);
-}
-
-JL_DLLEXPORT void jl_module_use_as(jl_task_t *ct, jl_module_t *to, jl_module_t *from, jl_sym_t *s, jl_sym_t *asname)
-{
-    module_import_(ct, to, from, asname, s, 0);
+    assert(m);
+    jl_sym_t *name = asname ? asname : import->name;
+    // TODO: this is a bit race-y with what error message we might print
+    jl_binding_t *b = jl_get_module_binding(m, name, 1);
+    jl_binding_partition_t *bpart = jl_get_binding_partition(b, ct->world_age);
```

Comment:
This appears to be an existing mistake in `module_import_`, but `ct->world_age` is not correct to use in this function (these functions apply to `jl_world_counter`, so they will operate incorrectly if they instead reference world_age), so we should make a separate issue to fix this before the v1.12 release

Reply:
I think we avoided this bug on master by forcing imports to go through `jl_toplevel_eval_in`, so we'd always be in the latest world:
https://github.com/JuliaLang/julia/blob/6d78a4ae91078cea43470f8cd6d895e0c9b3d5e7/src/toplevel.c#L1094-L1095

Also discovered I broke `import Package as M` while looking in to this, woops. At least we'll have a test for it now.